Source/JavaScriptCore/b3/B3ReduceStrength.cpp

```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
         return false;
     }
 
+    // For Op==BitOr or BitXor, turn any of these:
+    //      Op(Shl(x1, x2), ZShr(x1, 32|64-x2))
+    //      Op(Shl(x1, x2), ZShr(x1, Sub(32|64, x2)))
+    //      Op(ZShr(x1, 32|64-x2), Shl(x1, x2))
+    //      Op(ZShr(x1, Sub(32|64, x2)), Shl(x1, x2))
+    // Into this: RotL(x1, x2)
+    // And any of these:
+    //      Op(ZShr(x1, x2), Shl(x1, 32|64-x2))
+    //      Op(ZShr(x1, x2), Shl(x1, Sub(32|64, x2)))
+    //      Op(Shl(x1, 32|64-x2), ZShr(x1, x2))
+    //      Op(Shl(x1, Sub(32|64, x2)), ZShr(x1, x2))
+    // Into this: RotR(x1, x2)
+    // This pattern is widely used to represent bitwise rotation in languages that don't support it natively,
+    // and appears in hot loops in cryptography algorithms, so we should try to match it when possible.
+    bool handleBitRotation()
```

Comment:
nit: Maybe we can add `ASSERT(m_value->opcode() == BitOr || m_value->opcode() == BitXor);` at the beginning.

Reply:
I'll add this in the next push.